File: AEDS2 - Java/Verde/TP3/JAVA/TP03Q03 - Pilha com Alocação Sequencial em Java/Alfa.java
import java.util.*;
import java.lang.*;



class Jogador {
    private int id;
    private String nome;
    private int altura;
    private int peso;
    private String universidade;
    private int anoNascimento;
    private String cidadeNascimento;
    private String estadoNascimento;
    public Jogador(){}
    public Jogador(int id, String nome, int altura, int peso, String universidade, int anoNascimento, String cidadeNascimento, String estadoNascimento){
        setId(id);
        setNome(nome);
        setAltura(altura);
        setPeso(peso);
        setUniversidade(universidade);
        setAnoNascimento(anoNascimento);
        setCidadeNascimento(cidadeNascimento);
        setEstadoNascimento(estadoNascimento);
    }
    public int getId(){
        return id;
    }

    public String getNome() {
        return nome;
    }

    public int getAltura() {
        return altura;
    }

    public int getPeso() {
        return peso;
    }

    public int getAnoNascimento() {
        return anoNascimento;
    }

    public String getCidadeNascimento() {
        return cidadeNascimento;
    }

    public String getEstadoNascimento() {
        return estadoNascimento;
    }

    public String getUniversidade() {
        return universidade;
    }
    public void setId(int id){
        this.id = id;
    }

    public void setAltura(int altura) {
        this.altura = altura;
    }

    public void setAnoNascimento(int anoNascimento) {
        this.anoNascimento = anoNascimento;
    }
    public void setPeso(int peso) {
        this.peso = peso;
    }

    public void setCidadeNascimento(String cidadeNascimento) {
        cidadeNascimento = validar(cidadeNascimento);
        this.cidadeNascimento = cidadeNascimento;
    }

    public void setEstadoNascimento(String estadoNascimento) {
        estadoNascimento = validar(estadoNascimento);
        this.estadoNascimento = estadoNascimento;
    }

    public void setNome(String nome) {
        nome = validar(nome);
        this.nome = nome;
    }
    public void setUniversidade(String universidade) {
        universidade = validar(universidade);
        this.universidade = universidade;
    }
    public String validar(String str){
        if(str.isEmpty())
            str = "nao informado";
        return str;
    }
    public Jogador clone(){
        return new Jogador(getId(),getNome(),getAltura(),getPeso(),getUniversidade(),getAnoNascimento(),getCidadeNascimento(),getEstadoNascimento());
    }


    public void imprimir() {
        System.out.println("## " + getNome() + " ## " +getAltura() + " ## " + getPeso() + " ## " + getAnoNascimento() + " ## " + getUniversidade() + " ## " + getCidadeNascimento() + " ## "+ getEstadoNascimento() +" ##");
    }
    public void ler(String jogador){

        String[] dados = jogador.split(",", 8);
        setId(dados[0]!="" ?Integer.parseInt(dados[0]): -1);
        setNome(dados[1]);
        setAltura(Integer.parseInt(dados[2]));
        setPeso(Integer.parseInt(dados[3]));
        setUniversidade(dados[4]);
        setAnoNascimento(Integer.parseInt(dados[5]));
        setCidadeNascimento(dados[6]);
        setEstadoNascimento(dados[7]);
    }



}
/**
 * Lista estatica
 * @author Max do Val Machado
 * @version 2 01/2015
 */
class pilhaSequencial {
    private Jogador[] array;
    private int n;


    /**
     * Construtor da classe.
     */
    public pilhaSequencial() {
        this(3922);
    }


    /**
     * Construtor da classe.
     *
     * @param tamanho Tamanho da lista.
     */
    public pilhaSequencial(int tamanho) {
        array = new Jogador[tamanho];
        n = 0;
    }


    /**
     * Insere um elemento na primeira posicao da lista e move os demais
     * elementos para o fim da lista.
     *
     * @param x int elemento a ser inserido.
     * @throws Exception Se a lista estiver cheia.
     */
    public void inserirInicio(Jogador x) throws Exception {

        //validar insercao
        if (n >= array.length) {
            throw new Exception("Erro ao inserir! Lista cheia!");
        }

        //levar elementos para o fim do array
        for (int i = n; i > 0; i--) {
            array[i] = array[i - 1].clone();
        }

        array[0] = x.clone();
        n++;
    }



    /**
     * Remove um elemento da primeira posicao da lista e movimenta
     * os demais elementos para o inicio da mesma.
     *
     * @return resp int elemento a ser removido.
     * @throws Exception Se a lista estiver vazia.
     */
    public Jogador removerInicio() throws Exception {

        //validar remocao
        if (n == 0) {
            throw new Exception("Erro ao remover!");
        }

        Jogador resp = array[0].clone();
        n--;

        for (int i = 0; i < n; i++) {
            array[i] = array[i + 1].clone();
        }

        return resp;
    }


    /**
     * Mostra os elementos da lista separados por espacos.
     */
    public void mostrar() {
        for (int i = n - 1; i > -1; i--) {
            System.out.print("[" + (n - 1 - i) + "] ");
            array[i].imprimir();
        }
    }


    /**
     * Procura um elemento e retorna se ele existe.
     *
     * @param x int elemento a ser pesquisado.
     * @return <code>true</code> se o array existir,
     * <code>false</code> em caso contrario.
     */
    public boolean pesquisar(int x) {
        boolean retorno = false;
        for (int i = 0; i < n && retorno == false; i++) {
            retorno = (array[i].getId() == x);
            ;
        }
        return retorno;
    }

}
public class Alfa {
    private static pilhaSequencial pilha;
    private static Jogador [] jogadores;

    public static void operacao(String data) throws Exception {
        String[] entrada = data.split(" ",3);
        int posicao = 0;
        if (entrada.length > 1)
            posicao = Integer.parseInt(entrada[1]);
        Jogador retorno;
        switch (entrada[0]) {
            case "I":
                pilha.inserirInicio(jogadores[posicao]);
                break;
            case "R":
                retorno = pilha.removerInicio();
                MyIO.println("(R) " + retorno.getNome());
                break;
        }
    }
    /*public static void criarLog(){
       Arq.openWrite("806347_selecao.txt");
       Arq.println("806347\t" + tempo + "ms" + "\t" + cont + "comparacoes" + "\t" + contMov + "movimentacoes");
       Arq.close();
   }*/
    public static void leArquivo(Jogador[] jogadores){
        Arq.openRead("players.csv");
        Arq.readLine(); //remove a primeira linha
        for(int i = 0; i < 3922; i++){
            String data = Arq.readLine();
            jogadores[i] = new Jogador();
            jogadores[i].ler(data);
        }
    }
    //metodo para identificar a parada, FIM
    static boolean parada(String palavra, String parada){
        for(int i = 0; i < parada.length(); i++){
            if(palavra.charAt(i) != parada.charAt(i))
                return true;
        }
        return false;
    }



    public static void main(String[] args) throws Exception {
        int n = 3922;
        int realTam = 0;

        jogadores = new Jogador[n];

        leArquivo(jogadores);
        int posicao;
        String parada = MyIO.readString();
        pilha = new pilhaSequencial(n);

        while (parada(parada, "FIM")) {
            posicao = Integer.parseInt(parada);
            pilha.inserirInicio(jogadores[posicao]);
            parada = MyIO.readString();
        }

        int quant = MyIO.readInt();
        for (int i = 0; i < quant; i++) {
            parada = MyIO.readLine();
            operacao(parada);

        }
        pilha.mostrar();
    }
}

File: AEDS2 - Java/Verde/TP3/C/TP03Q04 - Fila Circular com Alocação Sequencial em C/Alfa.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>
#include <time.h>

typedef struct Jogador{

    int id;
    char nome[100];
    int altura;
    int peso;
    char universidade[100];
    int anoNascimento;
    char cidadeNascimento[100];
    char estadoNascimento[100];
} Jogador;

char* validar(char str[]){

    if(!str || strcmp(str, "\0") == 0){ //se o campo for vazio, nao informado eh inserido

        return "nao informado";

    }

    return str;
}

void clone(Jogador * jogador, Jogador * copia){

    copia->id = jogador->id;
    strcpy(copia->nome, jogador->nome);
    copia->altura = jogador->altura;
    copia->peso = jogador->peso;
    strcpy(copia->universidade, jogador->universidade);
    copia->anoNascimento = jogador->anoNascimento;
    strcpy(copia->cidadeNascimento, jogador->cidadeNascimento);
    strcpy(copia->estadoNascimento, jogador->estadoNascimento);

}


void imprimir(Jogador * jogador) {
    printf("## %s ## %d ## %d ## %d ## %s ## %s ## %s ##\n", jogador->nome, jogador->altura, jogador->peso,
           jogador->anoNascimento, jogador->universidade, jogador->cidadeNascimento, jogador->estadoNascimento);
}

void ler(char* str, Jogador* jogador){

    char* token = strsep(&str, ",");  // aqui estou separando a string, a linha inteira do arquivo, usando como criterio de separacao a virgula
    if(token != NULL){
        jogador->id = atoi(token);
    }
    else
        jogador->id = -1;
    token = strsep(&str, ",");
    strcpy( jogador->nome, validar(token)); // aqui estou chamando a funcao validar para os atributos do tipo char, se nao houver nada, o nao informado eh inserido

    token = strsep(&str, ",");
    jogador->altura = atoi(token);

    token = strsep(&str, ",");
    jogador->peso = atoi(token);

    token = strsep(&str, ",");
    strcpy( jogador->universidade, validar(token));

    token = strsep(&str, ",");
    jogador->anoNascimento = atoi(token);

    token = strsep(&str, ",");
    strcpy(jogador->cidadeNascimento, validar(token));

    token = strsep(&str, "\n"); // aqui o criterio de separacao foi diferente, como o final da linha acaba em \n , tive que elimina-lo para o print sair corretamente
    strcpy( jogador->estadoNascimento, validar(token));

}


void leArquivo(Jogador* jogadores[]){
    FILE *fp = NULL;
    fp = fopen("/tmp/players.csv", "r");
    char *dados;

    dados = (char*) malloc(sizeof(char) * 1000);


    fgets(dados, 1000, fp);

    for(int i = 0; i < 3922; i++){ // aqui eu coloquei todos os jogadores em um array de jogadores, em que cada iteracao eu instanciava um jogador novo em cada indice do array
        fgets(dados, 1000, fp);
        jogadores[i] = (Jogador *) malloc(sizeof(Jogador));

        ler(dados, jogadores[i]);
    }
    free(dados);

}
int encontrarMenor(Jogador *array[], int n, int j, int *menor, int *comp) {
    if (j < n) {


        (*comp)++; // Contagem de comparações
        if (strcmp(array[*menor]->nome, array[j]->nome) > 0) {
            (*menor) = j; // Atualiza o índice do menor elemento
        }

        (*menor) = encontrarMenor(array, n, j + 1, menor, comp);
    }
    return *menor;
}




/*void criarLog (double tempo, int comp, int mov){
    FILE *fp = NULL;
    fp = fopen("806347_selecaoRecursiva.txt", "w");
    fprintf(fp, "806347\t %.4fs \t %d comparacoes \t %d movimentacoes", tempo, comp, mov);
    fclose(fp);
}*/
bool parada(char *str){ // aqui eh a funcao que verifica o criterio de parada
    if(str[0] == 'F' && str[1] == 'I' && str[2] == 'M')
        return false;
    else
        return true;
}

/**
/"Classe Lista"
**/

/**
 * fila estatica
 * @author Max do Val Machado
 * @version 2 01/2015
 */

#define MAXTAM    6

Jogador* array[MAXTAM];   // Elementos da fila circular
int primeiro;	            // Remove do indice "primeiro".
int ultimo;                 // Insere no indice "ultimo".
int quant;


/**
 * Inicializacoes
 */
void start(){
    primeiro = ultimo = quant = 0;
}

/**
 * Remove um elemento da primeira posicao da fila e movimenta
 * os demais elementos para o primeiro da mesma.
 * @return resp int elemento a ser removido.
 * @Se a fila estiver vazia.
 */
Jogador* remover() {

    //validar remocao
    if (primeiro == ultimo) {
        printf("Erro ao remover!");
        exit(1);
    }
    //mostrar();
    Jogador* resp = (Jogador*) malloc(sizeof (Jogador));
    clone(array[primeiro], resp);
    primeiro = (primeiro + 1) % MAXTAM;


    return resp;
}


/**
 * Insere um elemento na ultima posicao da
 * @param x int elemento a ser inserido.
 * @Se a fila estiver cheia.
 */
void inserir(Jogador* x) {

    //validar insercao
    if (((ultimo + 1) % MAXTAM) == primeiro) {
        remover();


    }else if(quant<MAXTAM-1)
        quant++;
    array[ultimo] = (Jogador*) malloc(sizeof(Jogador));
    clone(x, array[ultimo]);
    ultimo = (ultimo + 1) % MAXTAM;
}

int Round(float number)
{
    return (number >= 0) ? (int)(number + 0.5) : (int)(number - 0.5);
}
//tira media das alturas de cada jogador na fila
int mediaAltura(){
    int soma = 0;

    for(int i = primeiro;i!= ultimo;i= ((i + 1) % MAXTAM)){

        soma+=array[i] -> altura;


    }

    float media = (float)soma/quant;

    return Round(media);

}




void diminuiQuant(){
    quant--;
}


void operacao(char* data, Jogador* jogadores []) {
    char *divisao[3];

    divisao[0] = strsep(&data, " ");
    divisao[1] = strsep(&data, " ");

    Jogador *jogador;
    int index = 0;

    if (divisao[1] != NULL)
        index = atoi(divisao[1]);

    if (strcmp("I", divisao[0]) == 0) {
        inserir(jogadores[index]);
        printf("%d\n", mediaAltura());

    } else if (strcmp("R", divisao[0]) == 0) {
        jogador = remover(index);
        printf("(R) %s\n", jogador->nome);
        diminuiQuant();
        free(jogador);


    }
}






int main(){
    int m = 3922;
    Jogador* jogadores[m];

    leArquivo(jogadores);


    int posicao;
    char frase [500];
    scanf("%s", frase);

    while(parada(frase)){ // aqui eu tive que ler a entrada como uma string e utilizar do metodo atoi, para transformar a string em um inteiro ->

        posicao = atoi(frase); // -> tive que fazer isso pois era o unico jeito de comparar numeros com o criterio de parada e utilizar como indice do array

        inserir(jogadores[posicao]);
        printf("%d\n", mediaAltura());

        scanf("%s", frase);

    }
    int a;
    int* quanti = (int*) malloc(sizeof(int));
    scanf("%d",quanti);

    int* i = (int*) malloc(sizeof(int));

    for(*i = 0; *i < *quanti; *i = *i + 1){
        char data [50];
        scanf(" %[^\n]", data);
        operacao(data, jogadores);
    }
    free(quanti);
    free(i);
    int cont = 0;
    for(int i = primeiro; i != ultimo; i = ((i + 1) % MAXTAM)){
        printf("[%d] ", cont++);
        imprimir(array[i]);
    }


}

